pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,210 @@
+	prefixLengths *counter.PrefixLengthCounter
+
+	// log is the logger for the FQDNDataServer
+	log *slog.Logger
+}
+
+var (
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+type PolicyUpdater interface {
+	// UpdatePolicyRules is used to update the current state of the policy rules at the
+	// gRPC server. These rules are sent to the standalone DNS proxy.
+	// This is currently being called whenever there is a policy regeneration event
+	// for an endpoint.
+	UpdatePolicyRules(map[identity.NumericIdentity]policy.SelectorPolicy, bool) error
+}
+
+// StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
+// SDP calls this method to subscribe to DNS policies
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+// Note: this method is left empty on purpose and will be update with the actual implementation in the future PRs for the standalone DNS proxy
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
+	stream.Send(&pb.PolicyState{RequestId: "test"})
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg messagehandler.DNSMessageHandler, port int, logger *slog.Logger) *FQDNDataServer {
+	return &FQDNDataServer{
+		port:                port,
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]string),
+		log:                 logger,
+		prefixLengths:       counter.DefaultPrefixLengthCounter(),
+	}
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIPMutex.Lock()
+	defer s.identityToIPMutex.Unlock()
+
+	ipNet := cidr.AsIPNet()
```

Comment:
Why are you using IPNet here to convert it to string, and not just `ipnet.Prefix`. `ipnet.Prefix` can be used as a map key, so no need to convert anything to a string.

Reply:
I removed the usage of `string` and storing the ipnet.

Reply:
Is there any reason you use `currentIdentityToIP map[identity.NumericIdentity][]net.IPNet` instead of `currentIdentityToIP map[identity.NumericIdentity][]netip.Prefix`?